Mathlib/Topology/MetricSpace/BundledFun.lean

```diff
@@ -0,0 +1,191 @@
+/-
+Copyright (c) 2025 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import Mathlib.Algebra.Order.Monoid.Defs
+import Mathlib.Topology.UniformSpace.Ultra.Basic
+
+/-!
+# Pseudometrics as bundled functions
+
+This file defines a pseudometric as a bundled function.
+This allows one to define a semilattice on them, and to construct families of pseudometrics.
+
+## Implementation notes
+
+The `PseudoMetricFun` works over any ordered commutative additive monoid,
+which allows one to talk about functions mapping into `ℝ`, `ℝ≥0`, `ℝ≥0∞`, etc.
+Proving that the pseudometric is nonnegative relies on a linear order,
+so the `PseudoMetricFun` is (overly) restricted to require a linear order in the codomain.
+The linear order also allows to define the supremum of two pseudometrics.
+
+-/
+
+variable {X R : Type*} [LinearOrderedAddCommMonoid R]
+
+variable (X R) in
+/-- A pseudometric as a bundled function. -/
+@[ext]
+structure PseudoMetricFun where
+  /-- The underlying binary function mapping into a linearly ordered additive monoid. -/
+  toFun : X → X → R
+  /-- A pseudometric must take identical elements to 0. -/
+  refl' x : toFun x x = 0
+  /-- A pseudometric must be symmetric. -/
+  symm' x y : toFun x y = toFun y x
+  /-- A pseudometric must respect the triangle inequality. -/
+  triangle' x y z : toFun x z ≤ toFun x y + toFun y z
+
+namespace PseudoMetricFun
+
+instance : FunLike (PseudoMetricFun X R) X (X → R) where
+  coe := PseudoMetricFun.toFun
+  coe_injective' := by
+    intro a
+    aesop
+
+variable (d : PseudoMetricFun X R)
+
+@[simp]
+lemma mk_apply (d : X → X → R) (refl symm triangle) (x y : X) :
+    PseudoMetricFun.mk d refl symm triangle x y = d x y :=
+  rfl
+
+@[simp]
+protected lemma refl (x : X) : d x x = 0 := d.refl' x
+protected lemma symm (x y : X) : d x y = d y x := d.symm' x y
+protected lemma triangle (x y z : X) : d x z ≤ d x y + d y z := d.triangle' x y z
+
+instance : LE (PseudoMetricFun X R) := ⟨fun d d' ↦ ∀ x y, d x y ≤ d' x y⟩
+
+@[simp]
+protected lemma le_iff_coe_le_coe (d d' : PseudoMetricFun X R) :
+    d ≤ d' ↔ (d : X → X → R) ≤ d' :=
+  Iff.rfl
```

Comment:
I would certainly make this simp and norm_cast the other way around. Also can `d`, `d'` be made implicit?